packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIteratorItem.tsx

```diff
@@ -72,27 +71,38 @@ export const SimpleFormIteratorItem = React.forwardRef(
         const sourceContext = useMemo(
             () => ({
                 getSource: (source: string) => {
-                    if (parentSourceContext) {
+                    if (!source) {
+                        // source can be empty for scalar values, e.g.
+                        // <ArrayInput source="tags" /> => SourceContext is "tags"
+                        //   <SimpleFormIterator> => SourceContext is "tags.0"
+                        //      <TextInput /> => use its parent's getSource so finalSource = "tags.0"
+                        //   </SimpleFormIterator>
+                        // </ArrayInput>
+                        return parentSourceContext.getSource(`${index}`);
+                    } else {
+                        // Normal input with source, e.g.
+                        // <ArrayInput source="orders" /> => SourceContext is "orders"
+                        //   <SimpleFormIterator> => SourceContext is "orders.0"
+                        //      <DateInput source="date" /> => use its parent's getSource so finalSource = "orders.0.date"
+                        //   </SimpleFormIterator>
+                        // </ArrayInput>
                         return parentSourceContext.getSource(
-                            source ? `${member}.${source}` : member
+                            `${index}.${source}`
                         );
                     }
-                    return source ? `${member}.${source}` : member;
                 },
                 getLabel: (source: string) => {
-                    // remove digits, e.g. 'book.authors.2.categories.3.identifier.name' => 'book.authors.categories.identifier.name'
```

Comment:
I think we should keep this comment

Reply:
We have a way better one now and you wrote it ;)